src/librustc/ty/layout.rs

```diff
@@ -907,18 +907,26 @@ impl<'tcx> LayoutCx<'tcx, TyCtxt<'tcx>> {
                         let count = (niche_variants.end().as_u32()
                             - niche_variants.start().as_u32()
                             + 1) as u128;
-                        // FIXME(#62691) use the largest niche across all fields,
-                        // not just the first one.
-                        for (field_index, &field) in variants[i].iter().enumerate() {
-                            let niche = match &field.largest_niche {
-                                Some(niche) => niche,
-                                _ => continue,
-                            };
-                            let (niche_start, niche_scalar) = match niche.reserve(self, count) {
-                                Some(pair) => pair,
-                                None => continue,
-                            };
-
+                        let mut niche_size = 0;
+                        if let Some((field_index, niche, niche_start, niche_scalar)) =
+                            variants[i].iter().enumerate().fold(None, |acc, (j, &field)| {
+                                let niche = match &field.largest_niche {
+                                    Some(niche) => niche,
+                                    _ => return acc,
+                                };
+                                let ns = niche.available(dl);
+                                if ns <= niche_size {
+                                    return acc;
+                                }
+                                match niche.reserve(self, count) {
+                                    Some(pair) => {
+                                        niche_size = ns;
+                                        Some((j, niche, pair.0, pair.1))
+                                    }
+                                    None => acc,
+                                }
```

Comment:
Consider sprinkling some comments here btw, and/or extracting functions :)

Reply:
Hopefully we can keep something close to the first commit, as it's more functional and way easier to explain as "find the largest niche among the fields" + "make sure it can fit our variant `count`".